docs/wave/shared_memory_barriers.rst

```diff
@@ -0,0 +1,119 @@
+How is shared memory barriers inserted in wave
+=============================================================
+
+We want to automatically insert shared-memory barriers so that read/write/atomics on the same LDS (shared memory) region are ordered correctly.
+To be more specific, cases like RAW (read after write), WAR (write after read), or touch memory via atomics, we must synchronize to avoid races and hangs. On gfx94, gfx95, we support basic shared memory barrier via `amdgpu.lds_barrier`; on gfx120x we prefer split barriers (Signal / Wait), split barriers are supported via lowering to `rocdl.s.wait_dscnt`, `rocdl.s.barrier.signal`, and `rocdl.s.barrier.wait`.
+
+Terminologies
+--------------------
+- Access types
+    * Read (READ)
+    * Write (WRITE)
+    * Atomic (READ_WRITE)
+    * GatherToLDS (Write but async)
+
+- Nested region op types
+    * Iterate
+    * Conditional
+
+**Note.** We will use producer to refer to an access type operator that takes ownership of a shared memory region for a period of time; consumer to refer to an access type operator that operates on the same shared memory region and therefore needs to synchronize utils the producer releases it.
+
+When do we need a barrier?
+--------------------
+- Basic shared memory barrier
+    * If access types of producer and consumer are different. -> insert a barrier in front of the consumer node.
+
+- GFX12 split barrier
+    * If access types of producer and consumer are different. -> insert a signal after the producer and a wait before the consumer
+
+- READ_WRITE is involved {acts like both a producer and a consumer}
+    * If it has a producer then this node will be treated like a consumer.
+    * If it has a consumer then this node will be treated like a producer.
+    * Either case, barriers are expected to be inserted with the logic described above.
+
+Visualization: add_shared_memory_barriers
+--------------------
+- Basic barrier
+
+.. image:: basic_barrier_vis.gif
+    :width: 400
+    :alt: Basic barrier GIF
+    :align: center
+
+The above gif is an visaul illustration for inserting shared memory barriers between producers and consumers.
+
+- Split barrier
+
+.. image:: split_barrier_vis.gif
+    :width: 400
+    :alt: Split barrier GIF
+    :align: center
+
+The above gif is an visaul illustration for inserting split barriers between producers and consumers.
+
+Heuristic: add_shared_memory_barriers
+--------------------
+The heuristic walks the graph in pre-order and proceeds as follows:
+
+0. Walks the graph in pre-order, node by node.
```

Comment:
This is a good first attempt at capturing the pass. Can we simplify this and maybe the pass as well?